### PR TITLE
[do-not-review] [fsdp] Allow bfloat16 mixed-precision reduction

### PR DESCRIPTION
**Pull request by xmfan**
Stacked PRs:
 * __->__#4434
 * #4431
 * #4436
 * #4435
 * #4433
 * #4432
 * #4430

--- --- ---

## Why

TorchTitan already forwards `training.mixed_precision_reduce` to PyTorch FSDP, and PyTorch supports both FP32 and BF16 reduction. TorchTitan restricts the configuration type to `float32`, preventing recipes from selecting the supported BF16 path.

BF16 reduction lowers collective bandwidth and memory traffic when the reduced precision is appropriate for the model policy.

## Change

Extend the public configuration literal to accept `bfloat16` while preserving `float32` as the default. No FSDP implementation or default behavior changes.

Add a deterministic two-GPU integration recipe that selects BF16 reduction and runs with a real process group for ten training steps.

## Testing

- `python -m pytest -q tests/unit_tests/cpu/test_config_manager.py -k bfloat16_mixed_precision_reduction`
- `python -m pytest -q tests/unit_tests/cpu/test_integration_test_definitions.py`
- `python -m tests.integration_tests.run_tests <output> --test_suite features --test_name fsdp_bf16_reduction --execution_mode real_pg --ngpu 2 --no-parallel`
- Integration result: 10/10 steps, loss decreases from 8.12030 to 3.98527

### CODE REVIEW

**Comment by xmfan** [comment]
should be addressed by [Support BF16 FSDP gradient reduction with BF16 fused optimizer state and FP32 master updates.](https://docs.google.com/spreadsheets/d/1ELMhvKn3i8xeD-y7D5T3upqS94eu9AJuswyWv9ek53U/edit#gid=810005&range=C3)
